chapter/8/big-data.md

```diff
@@ -1,31 +1,46 @@
+
+This chapter covers the original idea of MapReduce framework, split into two sections: programming model and execution model. For each section, we first introduce the original design for MapReduce and its limitations. Then we present follow-up models(e.g. FlumeJava) to either work around these limitations or other models (e.g. Dryad, Spark) that take alternative designs to circumvent inabilities of MapReduce. We also review declarative programming interfaces(Pig, Hive, SparkSQL) built on top of MapReduce frameworks to provide programming efficiency and optimization benefits. In the last section, we briefly outline the ecosystem of Hadoop and Spark.
+
+Outline  
+1. Programming Models
+  - 1.1 Data parallelism: MapReduce, FlumeJava, Dryad, Spark
+  - 1.2 Querying: Hive/HiveQL, Pig Latin, SparkSQL
+  - 1.3 Large-scale parallelism on Graph: BSP, GraphX  
+2. Execution Models
+  - 2.1 MapReduce execution model
+  - 2.2 Spark execution model
+  - 2.3 Hive execution model
+  - 2.4 SparkSQL execution model   
+3. Big Data Ecosystem:
+  - 3.1 Hadoop ecosystem
+  - 3.2 Spark ecosystem
+
 ## 1 Programming Models
 ### 1.1 Data parallelism
-*Data parallelism* is to run a single operation on different pieces of the data on different machines in parallel. Comparably, a sequential computation looks like *"for all elements in the dataset, do operation A"*, where dataset could be in the order of terabytes or petabytes aka. big data and one wants to scale up the processing. The challenges to do this sequential computation in a parallelized manner include how to abstract the different types of computations in a simple and correct way, how to distribute the data to hundreds/thousands of machines, how to handle failures and so on.
+*Data parallelism* is, given a dataset, the simultaneous execution on multiple machines or threads of the same function across groups of elements of a dataset. Data parallelism can also be thought of as a subset of SIMD ("single instruction, multiple data") execution, a class of parallel execution in Flynn's taxonomy. Comparably, one could think a sequential computation as *"for all elements in the dataset, do operation A"* on a single big dataset, whose size can reach to terabytes or petabytes. The challenges to doing this sequential computation in a parallelized manner include how to abstract the different types of computations in a simple and correct way, how to distribute the data to hundreds/thousands of machines or clusters, how to schedule tasks and handle failures and so on.
```

Comment:
"include how to abstract the different types of computations in a simple and correct way,"

I'm not sure what this means.